source/base/parameter_handler.cc

```diff
@@ -1697,16 +1697,53 @@ void ParameterHandler::parse_input (std::istream &input,
 
       if (!is_concatenated)
         {
-          scan_line (fully_concatenated_line, filename, current_logical_line_n);
+          // try to scan this line. if that fails, i.e., if scan_line throws
+          // an exception either because a parameter doesn't match its
+          // pattern or because an associated action throws an exception,
+          // then try to rewind the set of subsections to the same
+          // point where we were when the current function was called.
+          // this at least allows to read parameters from a predictable
+          // state, rather than leave the subsection stack in some
+          // unknown state.
+          //
+          // after unwinding the subsection stack, just re-throw the exception
+          try
+            {
+              scan_line (fully_concatenated_line, filename, current_logical_line_n);
+            }
+          catch (...)
+            {
+              while ((saved_path != subsection_path)
+                     &&
+                     (subsection_path.size() > 0))
+                leave_subsection ();
+
+              throw;
+            }
+
           fully_concatenated_line.clear();
         }
     }
 
   // While it does not make much sense for anyone to actually do this, allow
-  // the last line to end in a backslash.
+  // the last line to end in a backslash. to do do, we need to parse
```

Comment:
While not particularly important, my intent with the variable names in this section was that a 'concatenated line' referred to several lines that had been joined by `\`s. Hence it would be more consistent to write "To do this, we need to parse whatever concatenated line is left over.

Also, would it be possible to encapsulate this try-catch block with a lambda? It would be nice to not write the same thing in two places.